homeassistant/bootstrap.py

```diff
@@ -150,28 +159,32 @@ def from_config_dict(config: Dict[str, Any],
     _LOGGER.info("Home Assistant core initialized")
 
     # stage 0, load logging components
-    for component in components:
-        if component in LOGGING_COMPONENT:
-            hass.async_create_task(
-                async_setup_component(hass, component, config))
-
-    await hass.async_block_till_done()
+    await asyncio.gather(
+        *[
+            async_setup_component(hass, component, config)
+            for component in components
+            if component in LOGGING_COMPONENT
+        ], loop=hass.loop)
 
     # stage 1
-    for component in components:
-        if component in FIRST_INIT_COMPONENT:
-            hass.async_create_task(
-                async_setup_component(hass, component, config))
+    await asyncio.gather(
+        *[
+            async_setup_component(hass, component, config)
+            for component in components
+            if component in FIRST_INIT_COMPONENT
+        ], loop=hass.loop)
 
-    await hass.async_block_till_done()
+    # Ensure registries are loaded before stage 2.
+    await registry_load
```

Comment:
This is actually optional here, we don't actually need to await the registries loading since they're synchronised for loading anyhow.

Reply:
Are you sure  `hass.async_block_till_done()` can be removed? There are lots of usage in the component loading will not await the task finished.

Reply:
You're probably right, I did an audit of the logging components, they don't create any tasks so we don't have to block after step zero. 

Ideally the stage 1 components (or really all components) would not register any services or event listeners, or any other callables until they're ready to accept stuff coming in on those callables, otherwise there's a potential race-condition with any other stage 1 components calling those callables during the init process. Also they should not return from `async_setup` before all callables are registered.

The best way to arrange this for fast start-up if you did fire off a start-up task, would be to await on a `asyncio.Signal` in any callable method, and set the `Signal` once any start up task has progressed to the point where it can accept stuff arriving on the callables. This means anything coming in is queued awaiting startup, but nothing gets blocked unless it needs to call the component.

I couldn't be sure if the mqtt component does this, it connects to the mqtt server and it got a bit too complicated to see if anything needing this connection waited for it to be ready. 

Given this I have put in the wait again, but doing an audit of the stage one components to ensure this property is true would result in faster start-up times.